### PR TITLE
stop using a carry on FIFO queue to observe pods

### PR DESCRIPTION
**Pull request by deads2k**
`NewResyncableFIFO` is a carry on kubernetes in a difficult to read area.  We developed new watch tools and we should use this. This updates us to use the watchtools and eliminates a carry.

/assign @tnozicka

### CODE REVIEW

File: pkg/apps/strategy/support/lifecycle.go [hunk 1]
```diff
@@ -224,58 +226,93 @@ func (e *hookExecutor) executeExecNewPod(hook *appsv1.LifecycleHook, rc *corev1.
...
-				wg.Done()
-				break waitLoop
-			}
-			if !created {
-				fmt.Fprintf(e.out, "--> %s: Hook pod is already running ...\n", label)
+				return false, fmt.Errorf("%s", message)
+			case watch.Added, watch.Modified:
+				updatedPod = event.Object.(*corev1.Pod)
+			case watch.Deleted:
+				err := fmt.Errorf("%s: pod/%s[%s] unexpectedly deleted", label, pod.Name, pod.Namespace)
```
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/apps/strategy/support/lifecycle.go]: > nit: `pod namespace/name`

that makes it look like a resource string for oc that cannot be used.

**Comment by deads2k** [comment]
comments addressed

File: pkg/apps/strategy/support/lifecycle.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tnozicka) [reply pkg/apps/strategy/support/lifecycle.go]: that is workqueue label, a standard way workload controller log it. not saying it's ideal, just consistent.

I get your point about oc having a different syntax, not particularly in favor of the namespace `[]` syntax though.

File: pkg/apps/strategy/support/lifecycle.go [hunk 2]
```diff
@@ -224,58 +226,101 @@ func (e *hookExecutor) executeExecNewPod(hook *appsv1.LifecycleHook, rc *corev1.
...
+			return e.pods.Pods(pod.Namespace).Watch(options)
+		},
+	}
+	// make sure that the pod exists and wasn't deleted early
+	preconditionFunc := func(store cache.Store) (bool, error) {
+		_, exists, err := store.Get(&metav1.ObjectMeta{Namespace: pod.Namespace, Name: pod.Name})
+		if err != nil {
+			return true, err
+		}
+		if !exists {
```
Comment (tnozicka) [inline pkg/apps/strategy/support/lifecycle.go]: you could also verify uid, to avoid case of re-create, but that would be super rare

**Pull request merged**